homeassistant/components/fritz/common.py

```diff
@@ -0,0 +1,270 @@
+
+Device = namedtuple("Device", ["mac", "ip", "name"])
+
+
+CONFIG_SCHEMA = vol.Schema(
+    vol.All(
+        cv.deprecated(DOMAIN),
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_DEVICES): vol.All(
+                        cv.ensure_list,
+                        [
+                            vol.Schema(
+                                {
+                                    vol.Optional(CONF_HOST): cv.string,
+                                    vol.Optional(CONF_PORT): cv.port,
+                                    vol.Required(CONF_USERNAME): cv.string,
+                                    vol.Required(CONF_PASSWORD): cv.string,
+                                }
+                            )
+                        ],
+                    )
+                }
+            )
+        },
+    ),
+    extra=vol.ALLOW_EXTRA,
+)
+
+SERVICE_SCHEMA = vol.Schema({vol.Required(ATTR_HOST): cv.string})
+
+
+class FritzBoxTools:
+    """FrtizBoxTools class."""
+
+    def __init__(
+        self,
+        hass,
+        password,
+        username=DEFAULT_USERNAME,
+        host=DEFAULT_HOST,
+        port=DEFAULT_PORT,
+    ):
+        """Initialize FritzboxTools class."""
+        self._cancel_scan = None
+        self._device_info = None
+        self._devices: Dict[str, Any] = {}
+        self._unique_id = None
+        self.connection = None
+        self.error = None
+        self.fritzhosts = None
+        self.fritzstatus = None
+        self.ha_ip = None
+        self.hass = hass
+        self.host = host
+        self.password = password
+        self.port = port
+        self.success = None
+        self.username = username
+
+    async def async_setup(self):
+        """Wrap up FritzboxTools class setup."""
+        return await self.hass.async_add_executor_job(self.setup)
+
+    def setup(self):
+        """Set up FritzboxTools class."""
+        try:
+            self.connection = FritzConnection(
+                address=self.host,
+                port=self.port,
+                user=self.username,
+                password=self.password,
+                timeout=60.0,
+            )
+
+            self.fritzstatus = FritzStatus(fc=self.connection)
+            self._unique_id = self.connection.call_action("DeviceInfo:1", "GetInfo")[
+                "NewSerialNumber"
+            ]
+            self.fritzhosts = FritzHosts(fc=self.connection)
+            self._device_info = self._fetch_device_info()
+            self.success = True
+            self.error = False
+        except FritzConnectionException:
+            self.success = False
+            self.error = ERROR_CONNECTION_ERROR
+
+        self.ha_ip = get_local_ip()
+
+        self.scan_devices()
+
+        self._cancel_scan = async_track_time_interval(
+            self.hass, self.scan_devices, timedelta(seconds=TRACKER_SCAN_INTERVAL)
+        )
+
+        return self.success, self.error
```

Comment:
Since the error is always the same and self.success and self.error are mutually exclusive, it would be simpler to let it blow up here and wrap all calls in try blocks intead.

example


```

    try:
        await fritz_tools.async_setup()
    except FritzConnectionException:
        _LOGGER.error("Unable to setup FRITZ!Box Tools component")
        hass.async_create_task(
            hass.config_entries.flow.async_init(
                DOMAIN,
                context={"source": SOURCE_REAUTH},
                data=entry,
            )
        )
        return False
```

Reply:
Fixed